app/jobs/obtain_annotations_with_callback_job.rb

```diff
@@ -6,50 +6,31 @@ class ObtainAnnotationsWithCallbackJob < ApplicationJob
 
   def perform(project, filepath, annotator, options)
     doc_count = File.read(filepath).each_line.count
-
     prepare_progress_record(doc_count)
 
-    # for asynchronous protocol
-    doc_collection = DocCollection.new(project, annotator, @job.id, options)
+    doc_packer = DocPacker.new(annotator, encoding: options[:encoding])
 
     File.foreach(filepath) do |line|
-      docid = line.chomp.strip
-      doc = Doc.find(docid)
-      doc.set_ascii_body if options[:encoding] == 'ascii'
-
-      if doc_collection.filled_with?(doc)
-        begin
-          request_info = doc_collection.request_annotate
-          add_sliced_doc_exception_message_to_job(request_info, doc) if error_occured?(request_info)
-          update_job_items(annotator, doc_collection.docs.first, request_info.count)
+      doc_packer << line.chomp.strip
+    end
 
-        rescue StandardError, RestClient::RequestFailed => e
-          less_docs_message = 'Could not obtain annotations:'
-          many_docs_message = 'Could not obtain annotations for'
-          add_exception_message_to_job(doc_collection.docs, e, less_docs_message, many_docs_message)
-        ensure
-          doc_collection.clear
-        end
+    doc_packer.each do |hdocs, doc, error|
+      # Handle document slice error
+      if error
+        add_exception_message_to_job(doc, error)
+        next
       end
 
-      doc_collection << doc
+      update_job_items(annotator, doc, hdocs.length) if hdocs.any? { _1.key?(:span) }
 
-    rescue RuntimeError => e
-      less_docs_message = 'Runtime error:'
-      many_docs_message = 'Runtime error while processing'
-      add_exception_message_to_job(doc_collection.docs, e, less_docs_message, many_docs_message)
-    end
-
-    if doc_collection.rest?
-      begin
-        request_info = doc_collection.request_annotate
-        add_sliced_doc_exception_message_to_job(request_info, doc_collection.docs.first) if error_occured?(request_info)
-        update_job_items(annotator, doc_collection.docs.first, request_info.count)
-
-      rescue StandardError, RestClient::RequestFailed => e
-        less_docs_message = 'Could not obtain annotations:'
-        many_docs_message = 'Could not obtain annotations for'
-        add_exception_message_to_job(doc_collection.docs, e, less_docs_message, many_docs_message)
+      hdocs_to_request = annotator.single_doc_processing? ? hdocs.map { [_1] } : [hdocs]
+      hdocs_to_request.each do |hdoc|
+        begin
+          make_request(annotator, project, hdoc, options)
+        rescue StandardError, RestClient::RequestFailed => e
+          add_exception_message_to_job(annotator.single_doc_processing? ? hdoc.first : hdoc, e)
+          break if e.class == RestClient::InternalServerError
+        end
       end
```

Comment:
```suggestion
      if annotator.single_doc_processing?
         hdocs.each do | hdoc |
            begin
              make_request(annotator, project, hdoc, options)
            rescue StandardError, RestClient::RequestFailed => e
              add_exception_message_to_job(annotator.single_doc_processing? ? hdoc.first : hdoc, e)
              break if e.class == RestClient::InternalServerError
            end
         end            
      else
       begin
          make_request(annotator, project, hdocc, options)
        rescue StandardError, RestClient::RequestFailed => e
          add_exception_message_to_job(annotator.single_doc_processing? ? hdoc.first : hdoc, e)
          break if e.class == RestClient::InternalServerError
        end
      end
```

このようにif/elseで書いてしまった方がわかりやすいと思います。
さらに例外処理をループ用とそれ以外で見直せそうです。